config/daemonconfig/daemonconfig.go

```diff
@@ -168,7 +169,11 @@ func SupplementDaemonConfig(c DaemonConfig, imageID, snapshotID string,
 		// when repository is public.
 		keyChain := auth.GetRegistryKeyChain(registryHost, imageID, labels)
 		c.Supplement(registryHost, image.Repo, snapshotID, params)
-		c.FillAuth(keyChain)
+		if config.IsKeyringEnabled() && fn != nil {
+			fn(registryHost, keyChain)
```

Comment:
Can `fn` fail ?

Reply:
Yes, it depends on the caller `SupplementDaemonConfig` and the `fn` may fail.